From 8772e1264e019bae24e8799cb5719417bf4f6002 Mon Sep 17 00:00:00 2001
From: chenj <chenj@lemote.com>
Date: Thu, 1 Dec 2016 09:51:35 +0800
Subject: [PATCH 251/314] AOSCOS: platform: export psmouse::touchpad led device

[Mingcong Bai: Resolved a minor merge conflict in
drivers/platform/mips/Kconfig.]

Signed-off-by: Huacai Chen <chenhc@lemote.com>
Signed-off-by: Mingcong Bai <jeffbai@aosc.io>

Signed-off-by: chenj <chenj@lemote.com>
Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
---
 .../include/asm/mach-loongson64/ec_wpce775l.h |  7 +++
 drivers/platform/mips/Kconfig                 |  1 +
 drivers/platform/mips/lemote3a-laptop.c       | 44 +++++++++++++++++++
 3 files changed, 52 insertions(+)

diff --git a/arch/mips/include/asm/mach-loongson64/ec_wpce775l.h b/arch/mips/include/asm/mach-loongson64/ec_wpce775l.h
index fc60026bc674..f1c686ef7a54 100644
--- a/arch/mips/include/asm/mach-loongson64/ec_wpce775l.h
+++ b/arch/mips/include/asm/mach-loongson64/ec_wpce775l.h
@@ -127,6 +127,13 @@ enum
 };
 /* <<< End Backlight */
 
+#define INDEX_TOUCHPAD_ENABLE_LED	0x56
+enum
+{
+	TP_EN_LED_OFF,
+	TP_EN_LED_ON
+};
+
 /* >>> Read battery(BQ3060) index for ACPI 80h command */
 /*
  * The reported battery die temperature.
diff --git a/drivers/platform/mips/Kconfig b/drivers/platform/mips/Kconfig
index 7ef9a47556d5..a5a80e0a399f 100644
--- a/drivers/platform/mips/Kconfig
+++ b/drivers/platform/mips/Kconfig
@@ -49,6 +49,7 @@ config LEMOTE3A_LAPTOP
 	select INPUT_SPARSEKMAP
 	select INPUT_EVDEV
 	depends on INPUT
+	select LEDS_CLASS
 	default y
 	help
 	  Lemote Loongson-3A/2Gq family laptops driver.
diff --git a/drivers/platform/mips/lemote3a-laptop.c b/drivers/platform/mips/lemote3a-laptop.c
index a8d339a3e30b..b6f585b0f7d3 100644
--- a/drivers/platform/mips/lemote3a-laptop.c
+++ b/drivers/platform/mips/lemote3a-laptop.c
@@ -19,12 +19,14 @@
 #include <linux/fb.h>
 #include <linux/interrupt.h>
 #include <linux/pm.h>
+#include <linux/reboot.h>
 #include <linux/power_supply.h>
 #include <linux/input.h>
 #include <linux/input/sparse-keymap.h>
 #include <linux/jiffies.h>
 #include <linux/miscdevice.h>
 #include <linux/platform_device.h>
+#include <linux/leds.h>
 #include <asm/bootinfo.h>
 
 #include <ec_wpce775l.h>
@@ -218,6 +220,9 @@ static int lemote3a_bat_very_low_handler(int status);
 /* SCI device LID event handler */
 static int lemote3a_lid_handler(int status);
 
+static void lemote3a_tp_led_set(struct led_classdev *led_cdev,
+			       enum led_brightness brightness);
+
 /* Hotkey device init handler */
 static int lemote3a_hotkey_init(void);
 /* Hotkey device exit handler */
@@ -386,6 +391,23 @@ static const struct key_entry lemote3a_keymap[] =
 	{KE_END, 0 }
 };
 
+/* Touchpad EN/DIS led*/
+static struct led_classdev lemote3a_tp_led = {
+	.name = "psmouse::touchpad",
+	.brightness_set = lemote3a_tp_led_set,
+	.flags = LED_CORE_SUSPENDRESUME,
+};
+
+static int tp_led_shutdown_notify(struct notifier_block *unused1,
+			   unsigned long unused2, void *unused3)
+{
+	lemote3a_tp_led_set(&lemote3a_tp_led, LED_OFF);
+	return NOTIFY_DONE;
+}
+
+static struct notifier_block tp_led_nb = {
+	.notifier_call = tp_led_shutdown_notify,
+};
 
 static int wpce775l_probe(struct platform_device *dev)
 {
@@ -432,6 +454,13 @@ static int wpce775l_probe(struct platform_device *dev)
 	}
 	/* Register power supply END */
 
+	/* Touchpad enable/disable LED */
+	ret = led_classdev_register(NULL, &lemote3a_tp_led);
+	if (ret == 0)
+		register_reboot_notifier(&tp_led_nb);
+	else
+		goto fail_tp_led_register;
+
 	/* Hotkey device START */
 	ret = lemote3a_hotkey_init();
 	if (ret) {
@@ -467,6 +496,8 @@ static int wpce775l_probe(struct platform_device *dev)
 fail_sci_pci_driver_init:
 	lemote3a_hotkey_exit();
 fail_hotkey_init:
+	led_classdev_unregister(&lemote3a_tp_led);
+fail_tp_led_register:
 	power_supply_unregister(lemote3a_ac);
 fail_ac_power_supply_register:
 	power_supply_unregister(lemote3a_bat);
@@ -513,6 +544,10 @@ static void __exit lemote3a_laptop_exit(void)
 	sci_pci_driver_exit();
 	lemote3a_hotkey_exit();
 
+	/* Touchpad enable/disable LED */
+	unregister_reboot_notifier(&tp_led_nb);
+	led_classdev_unregister(&lemote3a_tp_led);
+
 	/* Power supply */
 	power_supply_unregister(lemote3a_ac);
 	power_supply_unregister(lemote3a_bat);
@@ -1136,6 +1171,15 @@ static int lemote3a_lid_handler(int status)
 	return 0;
 }
 
+/* Set touchpad en/dis led */
+static void lemote3a_tp_led_set(struct led_classdev *led_cdev,
+			       enum led_brightness brightness)
+{
+	int val = brightness ? TP_EN_LED_ON : TP_EN_LED_OFF;
+
+	ec_write(INDEX_TOUCHPAD_ENABLE_LED, val);
+}
+
 /* Hotkey device init handler */
 static int lemote3a_hotkey_init(void)
 {
-- 
2.48.1

